amethyst_tiles/src/pass.rs

```diff
@@ -0,0 +1,468 @@
+                .with_shaders(util::simple_shader_set(
+                    &shader_vertex,
+                    Some(&shader_fragment),
+                ))
+                .with_layout(&pipeline_layout)
+                .with_subpass(subpass)
+                .with_framebuffer_size(framebuffer_width, framebuffer_height)
+                .with_blend_targets(vec![pso::ColorBlendDesc(
+                    pso::ColorMask::ALL,
+                    pso::BlendState::PREMULTIPLIED_ALPHA,
+                )])
+                .with_depth_test(pso::DepthTest::On {
+                    fun: pso::Comparison::Less,
+                    write: false,
+                }),
+        )
+        .build(factory, None);
+
+    unsafe {
+        factory.destroy_shader_module(shader_vertex);
+        factory.destroy_shader_module(shader_fragment);
+    }
+
+    match pipes {
+        Err(e) => {
+            unsafe {
+                factory.device().destroy_pipeline_layout(pipeline_layout);
+            }
+            Err(e)
+        }
+        Ok(mut pipes) => Ok((pipes.remove(0), pipeline_layout)),
+    }
+}
+
+fn fetch_world_bounds<T: Tile, E: CoordinateEncoder>(
+    tile_map: &TileMap<T, E>,
+    world: &World,
+) -> (Vector3<u32>, Vector3<u32>) {
+    let camera_fetch = CameraGatherer::gather_camera_entity(world);
+    assert!(camera_fetch.is_some());
+    let camera_entity = camera_fetch.unwrap();
+
+    let (transforms, cameras) =
+        <(ReadStorage<'_, Transform>, ReadStorage<'_, Camera>)>::fetch(world);
+
+    //let camera_tile_id = entity_tile_ids.get(camera_entity).u wrap();
+    let camera_transform = transforms.get(camera_entity).unwrap();
+    let camera_translation = camera_transform.global_matrix().column(3).xyz();
+    let camera = cameras.get(camera_entity).unwrap();
+
+    assert!(camera.projection().as_orthographic().is_some());
+    let proj = camera.projection().as_orthographic().unwrap();
+
+    // Scale the camera view area
+    let world_top: Vector3<f32> = amethyst_core::math::convert(
+        camera_translation
+            + Vector3::new(
+                proj.left() * camera_transform.scale().x,
+                proj.top() * camera_transform.scale().y,
+                0.0,
+            ),
+    );
+
+    let world_bottom: Vector3<f32> = amethyst_core::math::convert(
+        camera_translation
+            + Vector3::new(
+                proj.right() * camera_transform.scale().x,
+                proj.bottom() * camera_transform.scale().y,
+                0.0,
+            ),
+    );
+
+    let one = Vector3::new(1, 1, 0);
+
+    let mut world_top_tile = tile_map
+        .to_tile(&world_top)
+        .map_or(Point3::new(0, 0, 0), |p| {
+            if p.x > 1 && p.y > 1 {
+                p - one
+            } else {
+                p
+            }
+        });
+
+    let mut world_bottom_tile =
+        tile_map
+            .to_tile(&world_bottom)
+            .map_or(Point3::from(*tile_map.dimensions()), |mut p| {
```

Comment:
I'm also getting [poor performance](https://wetransfer.com/downloads/70d1a823e033c863dc8dd6ac5019ce8320190810005408/46a51cf59600186567b5196d6454818d20190810005409/e55f23) with the example + debug build (I don't know what is causing that. I'll have come back to that some other day, getting too tired). While looking into the performance issue I noticed that it gets worse if the camera comes close to left or top borders of the map.

I'm not sure if this is really a problem at all and I don't understand the 3d maps so this might be wrong:

If I understand this will iterate tiles from ``(0, 0, 0)`` if the camera hits top or left borders (instead of ``(0, tile at camera top y, 0)`` when over the left border). That is potentially feeding render pass with more tiles than necessary. Also if camera bottom right > tilemap top left I think this function will return ``((0,0,0), (dimensions.x, dimensions.y, dimensions.z))`` when nothing should be in view.

Reply:
The performance issues is due to the large amount of multiplication occuring converting between tile and world space. Because of this, nalgebra just slows everything down a lot. 

I'll look into the other cases.

Reply:
I've now moved all the math into the shader, and it has resolved this problem.